types/wlr_virtual_keyboard.c

```diff
@@ -0,0 +1,208 @@
+		struct wl_resource *resource) {
+	virtual_keyboard_destroy_resource(resource);
+}
+
+static const struct zwp_virtual_keyboard_v1_interface virtual_keyboard_impl = {
+	.keymap = virtual_keyboard_keymap,
+	.key = virtual_keyboard_key,
+	.modifiers = virtual_keyboard_modifiers,
+	.destroy = virtual_keyboard_destroy,
+};
+
+static const struct zwp_virtual_keyboard_manager_v1_interface manager_impl;
+
+static struct wlr_virtual_keyboard_manager *manager_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwp_virtual_keyboard_manager_v1_interface, &manager_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void virtual_keyboard_manager_create_virtual_keyboard(
+		struct wl_client *client, struct wl_resource *resource,
+		struct wl_resource *seat,
+		uint32_t id) {
+	struct wlr_virtual_keyboard_manager *manager =
+		manager_from_resource(resource);
+
+	struct wlr_virtual_keyboard *virtual_keyboard = calloc(1,
+		sizeof(struct wlr_virtual_keyboard));
+	if (!virtual_keyboard) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_keyboard* keyboard = calloc(1, sizeof(struct wlr_keyboard));
+	if (!keyboard) {
+		wlr_log(L_ERROR, "Cannot allocate wlr_keyboard");
+		free(virtual_keyboard);
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wlr_keyboard_init(keyboard, &keyboard_impl);
+
+	struct wl_resource *keyboard_resource = wl_resource_create(client,
+		&zwp_virtual_keyboard_v1_interface, wl_resource_get_version(resource),
+		id);
+	if (!keyboard_resource) {
+		free(keyboard);
+		free(virtual_keyboard);
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(keyboard_resource, &virtual_keyboard_impl,
+		virtual_keyboard, virtual_keyboard_destroy_resource);
+
+	wlr_input_device_init(&virtual_keyboard->input_device,
+		WLR_INPUT_DEVICE_KEYBOARD, &input_device_impl, "virtual keyboard",
+		0xabcd, 0x1234);
+
+	struct wlr_seat_client *seat_client = wlr_seat_client_from_resource(seat);
+
+	virtual_keyboard->input_device.keyboard = keyboard;
+	virtual_keyboard->resource = keyboard_resource;
+	virtual_keyboard->seat = seat_client->seat;
+	wl_signal_init(&virtual_keyboard->events.destroy);
+	wlr_signal_emit_safe(&manager->events.new_virtual_keyboard, virtual_keyboard);
+}
+
+static const struct zwp_virtual_keyboard_manager_v1_interface manager_impl = {
+	.create_virtual_keyboard = virtual_keyboard_manager_create_virtual_keyboard,
+};
+
+static void virtual_keyboard_manager_bind(struct wl_client *client, void *data,
+		uint32_t version, uint32_t id) {
+	assert(client);
+	struct wlr_virtual_keyboard_manager *manager = data;
+
+	struct wl_resource *resource = wl_resource_create(client,
+		&zwp_virtual_keyboard_manager_v1_interface, version, id);
+
+	if (!resource) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(resource, &manager_impl, manager, NULL);
+	manager->resource = resource;
```

Comment:
If two clients bind to the global, `manager->resource` will be overwritten. You should use a list and `wl_resource_get_link` instead.